File: src/claude_cowork_sync/deploy.py
"""Atomic deployment helpers for merged profiles."""

from __future__ import annotations

import logging
from datetime import datetime, timezone
from pathlib import Path

logger = logging.getLogger(__name__)


def atomic_swap_profile(live_profile: Path, merged_profile: Path, backup_parent: Path) -> Path:
    """Atomically swaps live profile with merged profile, preserving backup."""

    if not live_profile.exists():
        message = f"Live profile not found: {live_profile}"
        logger.error(message)
        raise FileNotFoundError(message)
    if not merged_profile.exists():
        message = f"Merged profile not found: {merged_profile}"
        logger.error(message)
        raise FileNotFoundError(message)
    backup_parent.mkdir(parents=True, exist_ok=True)
    timestamp = datetime.now(tz=timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    backup_path = backup_parent / f"{live_profile.name}.backup.{timestamp}"
    logger.info("Moving live profile to backup: %s", backup_path)
    live_profile.rename(backup_path)
    logger.info("Promoting merged profile to live path: %s", live_profile)
    merged_profile.rename(live_profile)
    return backup_path

File: src/deploy.rs
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{bail, Context, Result};
use chrono::Utc;
use log::info;

pub fn atomic_swap_profile(
    live_profile: &Path,
    merged_profile: &Path,
    backup_parent: &Path,
) -> Result<PathBuf> {
    if !live_profile.exists() {
        bail!("Live profile not found: {}", live_profile.display());
    }
    if !merged_profile.exists() {
        bail!("Merged profile not found: {}", merged_profile.display());
    }

    fs::create_dir_all(backup_parent)
        .with_context(|| format!("Failed to create backup parent {}", backup_parent.display()))?;

    let timestamp = Utc::now().format("%Y%m%dT%H%M%SZ").to_string();
    let backup_path = backup_parent.join(format!(
        "{}.backup.{}",
        live_profile
            .file_name()
            .and_then(|name| name.to_str())
            .unwrap_or("Claude"),
        timestamp
    ));

    info!(
        "Moving live profile {} to backup {}",
        live_profile.display(),
        backup_path.display()
    );
    fs::rename(live_profile, &backup_path).with_context(|| {
        format!(
            "Failed to move live profile {} to {}",
            live_profile.display(),
            backup_path.display()
        )
    })?;

    info!(
        "Promoting merged profile {} to live path {}",
        merged_profile.display(),
        live_profile.display()
    );
    fs::rename(merged_profile, live_profile).with_context(|| {
        format!(
            "Failed to promote merged profile {} to {}",
            merged_profile.display(),
            live_profile.display()
        )
    })?;

    Ok(backup_path)
}
